internal/torcx/remote.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2018 CoreOS Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package torcx
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	errNilRemote          = errors.New("nil Remote")
+	errEmptyUsrMountpoint = errors.New("empty USR mountpoint")
+	errEmptyTemplateURL   = errors.New("empty remote URL template")
+)
+
+// evaluateURL evaluates the URL template for a remote
+// and performs variables substitution sourcing values from
+// `/etc/os-release`.
+func (r *Remote) evaluateURL(usrMountpoint string) (string, error) {
+	if r == nil {
+		return "", errNilRemote
+	}
+	if usrMountpoint == "" {
+		return "", errEmptyUsrMountpoint
+	}
+	if r.TemplateURL == "" {
+		return "", errEmptyTemplateURL
+	}
+
+	vars, any := needSubstitution(r.TemplateURL)
+	if !any {
+		return r.TemplateURL, nil
+	}
+
+	fp, err := os.Open(OsReleasePath)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to open %s", OsReleasePath)
+	}
+	defer fp.Close()
+	osMeta, err := parseOsRelease(fp)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to parse %s", OsReleasePath)
+	}
+	osMeta["COREOS_USR"] = usrMountpoint
+
+	url := r.TemplateURL
+	// TODO(lucab): this is suboptimal and repetitive, but I'm not sure about a compact yet idiomatic way.
```

Comment:
I can understand if you don't want to add a dependency on a small unknown project, but I do think https://github.com/euank/gotmpl would be less repetitive feeling.

I think it would just end up as:

```go
_, err := gotmpl.TemplateString(r.TemplateURL, gotmpl.MapLookup(map[string]string{}))
if err == nil {
   return r.TemplateURL, nil // no variables
}
if _, ok := err.(gotmpl.UnresolvedVariableError); !ok {
    return err // something wrong besides needing vars
}

return gotmpl.TemplateString(r.TemplateURL, gotmpl.MapLookup(osMeta))
```

It also has the benefit of allowing users to escape `$` characters if that's desirable, though realistically few urls will contain them anyways.

If you don't want to pull in a dependency that can do simple substitution, I do think this code would be shorter and more readable if it were generic (e.g. split into a pkg and just templated the k/v of a map, didn't go through the trouble of only substitution detected variables, etc).

Reply:
You are right, it's silly to keep this generic logic in here.